### DISCUSSION TITLE
Missing generic constraint for nullable types

### DISCUSSION THREAD

**Discussion by scalablecory** [dotnet/csharplang#2840] [Q&A]
_From @TobiasvdVen on Thursday, September 26, 2019 9:06:11 AM_

During our migration to .NET Core 3.0 we encountered a problem involving generic functions that might return null. See a minimal example below which generates the following warning:

`A null literal introduces a null value when 'T' is a non-nullable reference type. (CS8654)`

``` C#
public class Example
{
    public T GetSomething<T>() where T : class?
    {
        return null;
    }
}
```
[This](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/generics/constraints-on-type-parameters) page lists a generic constraint _notnull_, but not an equivalent for the opposite.

[This](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-8.0/nullable-reference-types-specification) page mentions a generic constraint _class?_. However, the documentation states that this denotes "possibly nullable reference type". Thus, a warning is generated when attempting to return null, as non-nullable reference types are still allowed by the constraint.

Is there a constraint that we have missed which enforces a generic parameter to be _nullable_, or will this be added in the future?

EDIT:

We found [this](https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/) page which describes _The issue with T_?. Is there a recommended method to implement a generic method that needs to be able to return null without being able to fully constrain this?

_Copied from original issue: dotnet/core#3476_

**Comment by scalablecory** [comment]
_From @scalablecory on Thursday, September 26, 2019 6:21:56 PM_

Why do you need it to be a constraint? Can you put it on the return type?

```c#
public class Example
{
    public T? GetSomething<T>() where T : class
    {
        return null;
    }
}
```

**Comment by scalablecory** [comment]
_From @TobiasvdVen on Friday, September 27, 2019 7:17:44 AM_

> Why do you need it to be a constraint? Can you put it on the return type?

Interestingly, we thought we had tried that, but apparently only with the _class?_ constraint (not _class_), which seems to cause a compiler error. This does work for most of our purposes.

We do also have a case where we wish the generic function to accept both nullable structs and nullable reference types. Based on [this](https://devblogs.microsoft.com/dotnet/try-out-nullable-reference-types/) page under _The issue with T?_ heading, however, it seems that this is not possible pragmatically due to underlying differences between int? types and reference types marked as nullable. If that is truly the case, then this issue can probably be closed and we will have to find a different method of achieving the same result as our current implementation.

**Comment by canton7** [comment]
There's an attribute for this. `[return: MaybeNull]` on the method, I think? 

> We do also have a case where we wish the generic function to accept both nullable structs and nullable reference types

Isn't that what happens by default? [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKAMQB2ArgDYkTAlwAEcBFVuuKAzNetQMLUDeu1A9mxQAWagBU4sABQBKfoL45BK6gFkAPAEsCMAHzSWsgNwLV67boD8B4mRNnVmlBgAMBgEQAzAPY+PDsrmFi6uNtJ2JIEqAL6OjqzsYpriBuLUMAAWOgDmsryOcThFQA).

Or do you mean that given a `T`, people can still pass in a nullable `T`? If so, use `[AllowNull]`. [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwD4AEBMBGAsAKBQBgAIUMA6AEQEsIBzAOwHtZKBjKUgYQYBM4BBOhAA2ATyiUoAblwBiOgFchQiMCFxCcQari5cKAMzE0hDoQDeuQleKGUAFkIAVOLAAUASkvWLOa38IAsgA8lHQwAHyu+u7Svv5WwaEwAPyRCkoxXvHBJPhpikKZcVYAvlmE5QbEDsGOkQDafEoMAO4AcgUAuk6EMAAWoTTu5uVlOGNAA===).

**Comment by sharwell** [comment]
I've found several cases where I wanted to use `where T : defaultable`. For example, this constraint would resolve dotnet/roslyn#38260.
